### ISSUE TITLE
Coop + Randomizer. Only works first launch, then fails until reboot.

### ISSUE THREAD

**Issue by ghost** [Cloudef/ModEngine2#9]
A few of my friends conned me into buying this game with the intent of playing together with randomizer.

So far I've been able to get this combination working however only on the initial run off a fresh boot and again only after a reboot.

Some possibly (ir)relevant details:

#### The guide is inaccurate and unclear.

Following it exactly never worked properly. NPCs will not have random gear and AI will be broken and won't patrol or do anything until attacked. What managed to make it work was to copy the `regulation.bin` from `randomizer` into the `mod` folder. I also created multiple config files because I do not know which one really is used as having to reboot after every launch is incredibly tedious. `seemlesscoopsettings.ini` (also differently named) now goes into `mods`, not `SeemlessCoop/cooppassword.ini`.

There's also a copy of modengine2 in `randomizer/diste/ModEngine/`. Should these be replaced with this fork? Safe to remove as they're not read? I don't know. I replaced mine with seemingly no effect but again I can't say for sure since I've tried so many different things I don't know what truly works and what does not.

#### Config files.

I was never able to get the mods to function until I added a `config_eldenring.toml` to the `game/` folder.

It's contents are for all intents and purposes identical:

```
[modengine]
debug = false

[extension.mod_loader]
enabled = true

loose_params = false

mods = [
    { enabled = true, name = "default", path = "randomizer" }
]

[extension.scylla_hide]
enabled = false
```

Does this help? Maybe...?

#### Zombie processes.

The game fails to exit cleanly, needing to be killed (outside of Steam's "stop") since it leaves behind most of wine's `*.exe`s which `winedevice.exe` needing multiple `pkill -9` to actually destroy the zombie process. No variant of wine exits properly, not GE, TK, proton exp, proton stable or proton next. I don't know if this is an issue caused by a mod or this fork. Apologies if it's not relevant however this doesn't happen with a clean vanilla ER.

#### Launch executable.

Every Windows specific guide says to use the me2 batch file (`launchmod_eldenring.bat`). Is there any reason we too are not running the me2 executable instead of vanilla `eldenring.exe`?

The batch files contents seems to be telling the game to load for ER and points it to the config file.
`.\modengine2_launcher.exe -t er -c .\config_eldenring.toml`

#### Windows behaves differently/unpredictable.

The most frustrating part is that the windows community seems to have a slew of different methods to make this work. For example, one friend had many issues and had to replicate my setup/install while another friend installed it completely wrong yet works flawlessly first try. Friend 1 and I both got `?EventTextForMap?` which according to the randomizer mod author is a "failsafe in the randomizer when it's not properly installed, meaning it only detected some of the files it needs.". Friend 2 never saw this despite his install being "broken" (and we had him stream to prove the issue was indeed not present.).

So, with all that said, I don't know. I'm hoping you or someone else reads this and can provide a definitive solution. I'll update should I figure something out.

**Comment by Cloudef** [comment]
The whole point of this fork is that the windows methods do not work on linux as it uses the detours lib from MS which is not compatible with wine. I have not tried this fork with latest elden ring patches so I do not know if it still works, sorry. It might be enough to just rebase this fork against upstread modengine2 to make it work however if elden ring updates are the problem. I do not play elden ring much anymore so I haven't kept this fork up-to-date.

**Comment by ghost** [comment]
Likely the new patch causing issues but it's strange that it works sometimes, usually on a fresh boot (though seems more like 65%-ish now).

Tried to rebase myself but failed.

```
Auto-merging src/modengine/ext/mod_loader/mod_loader_extension.cpp
CONFLICT (content): Merge conflict in src/modengine/ext/mod_loader/mod_loader_extension.cpp
error: could not apply 04005bd... temporary hardcoded offsets for elden ring
hint: Resolve all conflicts manually, mark them as resolved with
hint: "git add/rm <conflicted_files>", then run "git rebase --continue".
hint: You can instead skip this commit: run "git rebase --skip".
hint: To abort and get back to the state before "git rebase", run "git rebase --abort".
Could not apply 04005bd... temporary hardcoded offsets for elden ring
```

04005bd

edit: I think I correctly edited the file. Am I correct in assuming this needs to be compiled on windows?

**Comment by Cloudef** [comment]
If you send PR the github actions will compile it. I did not take the time to port this to mingw, unlike I did with the param randomizer https://github.com/Cloudef/elden_ring_param_randomiser

It could be also that since your issue is non deterministic, that it may be a race condition. Probably can be fixed by changing the load delay in elden mod loader https://github.com/Cloudef/ModEngine2/issues/1#issuecomment-1166694429

**Comment by Cloudef** [comment]
Also I think your rebase went slightly wrong: https://github.com/Cloudef/ModEngine2/commit/5696ed5852f13264027be32983e9eb86571b0f6f should be the top commit

**Comment by Cloudef** [comment]
Checking modengine2 upstream repo, there's only changes on main branch and they are DS3 related. So I think the issue isnt there, it's most likely the race in elden mod loader.

I rebased the code to `main` branch anyways, however it seems to not compile anymore due to some `vcpkg` shenanigan 🙁
https://github.com/Cloudef/ModEngine2/actions/runs/3586818931/jobs/6036507007#step:9:132
Sadly don't have time to port to mingw / nix right now.

**Comment by ghost** [comment]
I messed around with load times, from 0 to 7000. Didn't really seem to make a difference at all.

As for the race issue, that's above my pay grade I'm afraid.

Either way, you've already done more than I can ask for. Thanks for your time.

**Comment by Cloudef** [comment]
One more thing you may want to try is rename the .dll files in the mods folder so that the load order changes.
For example in the release README I had this
```
    ├── mods
    │   ├── lua.dll
    │   ├── modengine2.dll
    │   └── seamless_coop.dll
```
elden_ring_seamless_coop.dll was renamed to seamless_coop.dll to make sure it loads after modengine2.dll

EDIT:
Also reading your first post again. This modengine fork does not read config_eldenring.toml at all
>The modified modengine2 by default looks for config file in ELDEN RING/Game/modengine.toml, you can change this with MODENGINE_CONFIG environment variable if needed.

This is probably reason you had to copy regulation.bin as well 

>There's also a copy of modengine2 in randomizer/diste/ModEngine/. Should these be replaced with this fork? Safe to remove as they're not read?

That's bundled modengine2 with the randomizer, it can be removed

I'll suggest you read this again: https://github.com/Cloudef/ModEngine2/releases/tag/2.0.0.1-proton-v3
And especially check the file tree at the end as that's randomizer + seamless coop setup, everything there still should apply except for the seamless co-op password file which indeed should go into same directory as the dll nowadays.

As for the zombie processes, that I do not know why it happens, but you can use `wineserver -k` to kill all wine stuff. When using proton you can call it with protontricks (you need to know the gameid), IIRC something like:
`protontricks GAMEID -c wineserver -k`

**Comment by ghost** [comment]
> Also reading your first post again. This modengine fork does not read config_eldenring.toml at all

That's what I assumed but since everything worked that immediate launch once adding that file, I wasn't sure. It was a "maybe??" desperation thing.

> That's bundled modengine2 with the randomizer, it can be removed

Good to know.

> elden_ring_seamless_coop.dll was renamed to seamless_coop.dll to make sure it loads after modengine2.dll

I had tried delcaring load order it EldenRingLoader's config file (`mod_loader_config.ini`):

```
[loadorder]
lua.dll = 1
modengine2.dll = 2
elden_ring_seamless_coop.dll = 3
```

Seems this didn't work as intended though.

> As for the zombie processes, that I do not know why it happens

That's okay. I already scripted this but it's still a pain.

I'll try manually renaming the dlls and see if that helps.
